### DISCUSSION TITLE
Metadata Caching Problem

### DISCUSSION THREAD

**Discussion by sw4rty** [microsoft/PowerPlatform-DataverseServiceClient#274] [General]
Hi together,

we are using the Dataverse Client in an AppService to communicate with a Dynamics365 instance.
We have performance issues regarding the Entity- and Attribute-Metadata.

For all CRUD-Operation we have a lots of metadata calls.

Here is one example, we have like 13 calls before the create call.
![image](https://user-images.githubusercontent.com/105372985/167876914-c3d8369a-f62b-4bf2-ba19-1a2334afe005.png)

I tired to change from a scoped ServiceClient object to the Singleton/Clone ServiceClient approach (https://github.com/microsoft/PowerPlatform-DataverseServiceClient/issues/78#issuecomment-745414318) to fixed the problem.

But it didn't help because the metadata cache is not cloned.

Does anyone have an idea to improve our performance?

Regards
Oliver

**Comment by MattB-msft** [comment]
Yep,
This is a known issue / optimization we are working on improving.

The translation of ObjectModel to WebAPI is unavoidably heavy.
Server side ( where we go from WebAPI back to Object Model ) is supported by heavy caching, but is otherwise pretty much identical to what the client does.

If you set .UseWebAPI = false  You will see all that disappear.
I will try this as a bug here until we get the caching fix pushed to the client.

Thanks for your report!

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

From: ***@***.***>
Sent: Wednesday, May 11, 2022 8:02 AM
To: ***@***.***>
Cc: ***@***.***>
Subject: [microsoft/PowerPlatform-DataverseServiceClient] Metadata Caching Problem (Discussion #274)


Hi together,

we are using the Dataverse Client in an AppService to communicate with a Dynamics365 instance.
We have performance issues regarding the Entity- and Attribute-Metadata.

For all CRUD-Operation we have a lots of metadata calls.

Here is one example, we have like 13 calls before the create call.
[image]<https://user-images.githubusercontent.com/105372985/167876914-c3d8369a-f62b-4bf2-ba19-1a2334afe005.png>

I tired to change from a scoped ServiceClient object to the Singleton/Clone ServiceClient approach (#78 (comment)<https://github.com/microsoft/PowerPlatform-DataverseServiceClient/issues/78#issuecomment-745414318>) to fixed the problem.

But it didn't help because the metadata cache is not cloned.

Does anyone have an idea to improve our performance?

Regards
Oliver
